satellite/overlay/nodeselectioncache_test.go

```diff
@@ -0,0 +1,273 @@
+		addr := subnet + ".3:8080"
+		n := overlay.NodeCheckInInfo{
+			NodeID: storj.NodeID{byte(i)},
+			Address: &pb.NodeAddress{
+				Address:   addr,
+				Transport: pb.NodeTransport_TCP_TLS_GRPC,
+			},
+			LastNet:    subnet,
+			LastIPPort: addr,
+			IsUp:       true,
+			Capacity: &pb.NodeCapacity{
+				FreeDisk:      200 * memory.MiB.Int64(),
+				FreeBandwidth: 1 * memory.TB.Int64(),
+			},
+			Version: &pb.NodeVersion{
+				Version:    "v1.1.0",
+				CommitHash: "",
+				Timestamp:  time.Time{},
+				Release:    true,
+			},
+		}
+		err := db.UpdateCheckIn(ctx, n, time.Now().UTC(), nodeCfg)
+		require.NoError(t, err)
+	}
+}
+
+type mockdb struct {
+	mu        sync.Mutex
+	callCount int
+}
+
+func (m *mockdb) SelectAllStorageNodesUpload(ctx context.Context, selectionCfg overlay.NodeSelectionConfig) (reputable, new []*overlay.SelectedNode, err error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	sync2.Sleep(ctx, 500*time.Millisecond)
+	m.callCount++
+	return []*overlay.SelectedNode{}, []*overlay.SelectedNode{}, nil
+}
+func TestRefreshConcurrent(t *testing.T) {
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	// concurrent cache.refresh with high staleness, where high staleness means the
+	// cache should only be refreshed the first time we call cache.refresh (via cache.Init)
+	mockDB := mockdb{}
+	cache := overlay.NewNodeSelectionCache(zap.NewNop(),
+		&mockDB,
+		highStaleness,
+		nodeCfg,
+	)
+
+	var group errgroup.Group
+	group.Go(func() error {
+		return cache.Init(ctx)
+	})
+	group.Go(func() error {
+		return cache.Init(ctx)
+	})
+	err := group.Wait()
+	require.NoError(t, err)
+
+	require.Equal(t, 1, mockDB.callCount)
+
+	// concurrent cache.refresh (called via cache.Init) with low staleness, where low staleness
+	// means that the cache will refresh *every time* cache.refresh is called
+	mockDB = mockdb{}
+	cache = overlay.NewNodeSelectionCache(zap.NewNop(),
+		&mockDB,
+		lowStaleness,
+		nodeCfg,
+	)
+	group.Go(func() error {
+		return cache.Init(ctx)
+	})
+	group.Go(func() error {
+		return cache.Init(ctx)
+	})
+	err = group.Wait()
+	require.NoError(t, err)
+
+	require.Equal(t, 2, mockDB.callCount)
+}
+
+func TestGetNode(t *testing.T) {
+	satellitedbtest.Run(t, func(ctx *testcontext.Context, t *testing.T, db satellite.DB) {
+		var nodeCfg = overlay.NodeSelectionConfig{
+			AuditCount:       0,
+			UptimeCount:      0,
+			NewNodeFraction:  0.2,
+			MinimumVersion:   "v1.0.0",
+			OnlineWindow:     4 * time.Hour,
+			DistinctIP:       true,
+			MinimumDiskSpace: 100 * memory.MiB,
+		}
+		cache := overlay.NewNodeSelectionCache(zap.NewNop(),
+			db.OverlayCache(),
+			lowStaleness,
+			nodeCfg,
+		)
+		// the cache should have no nodes to start
+		reputable, new := cache.Size()
+		require.Equal(t, 0, reputable)
+		require.Equal(t, 0, new)
+
+		// add some nodes to the database
+		const nodeCount = 4
+		addNodesToNodesTable(ctx, t, db.OverlayCache(), nodeCount)
+
+		// confirm cache.GetNodes returns the correct nodes
```

Comment:
suggestion for future PR: there's a lot of existing tests around correct selection in satellite/overlay/selection_test.go and TestRandomizedSelection in satellite/overlay/service_test.go -- I think those will eventually need to be repurposed to test selection from this cache.

Reply:
when we plug this cache into the code to use with uploads, we will update tests